File: tests/Gantt/apireview/mainwindow.cpp
/****************************************************************************
**
** This file is part of the KD Chart library.
**
** SPDX-FileCopyrightText: 2001 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
**
** SPDX-License-Identifier: MIT
**
****************************************************************************/

#include "mainwindow.h"

#include "entrydelegate.h"
#include "entrydialog.h"
#include "ui_mainwindow.h"

#include <KDGanttConstraintModel>
#include <KDGanttDateTimeGrid>
#include <KDGanttGraphicsView>
#include <KDGanttLegend>
#include <QAbstractItemView>
#include <QDebug>
#include <QHeaderView>
#include <QLabel>
#include <QPointer>
#include <QStandardItem>
#include <QStandardItemModel>
#include <QTreeView>

MainWindow::MainWindow(QWidget *parent, Qt::WindowFlags flags)
    : QMainWindow(parent, flags)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    initModel();
    initActions();
    initItemDelegate();
    initGrid();

    auto *leftView = qobject_cast<QTreeView *>(ui->ganttView->leftView());
    Q_ASSERT(leftView);
    leftView->setColumnHidden(1, true);
    leftView->setColumnHidden(2, true);
    leftView->setColumnHidden(3, true);
    leftView->setColumnHidden(4, true);
    leftView->setColumnHidden(5, true);
    leftView->header()->setStretchLastSection(true);

    connect(ui->ganttView->leftView(), &QAbstractItemView::customContextMenuRequested,
            this, &MainWindow::showContextMenu);
    connect(ui->ganttView->selectionModel(), &QItemSelectionModel::selectionChanged,
            this, &MainWindow::enableActions);

    connect(ui->ganttView->graphicsView(), &KDGantt::GraphicsView::clicked,
            this, &MainWindow::slotClicked);
    connect(ui->ganttView->graphicsView(), &KDGantt::GraphicsView::qrealClicked,
            this, &MainWindow::slotDoubleClicked);
}

void MainWindow::initModel()
{
    model = new QStandardItemModel(0, 6, this);
    model->setHeaderData(0, Qt::Horizontal, tr("Task"));
    ui->ganttView->setModel(model);

    auto *l = new KDGantt::Legend;
    l->setWindowTitle(tr("Legend"));
    l->show();
    l->setModel(model);

    constraintModel = new KDGantt::ConstraintModel(this);
    ui->ganttView->setConstraintModel(constraintModel);
}

void MainWindow::initActions()
{
    newEntryAction = new QAction(tr("New entry"), this);
    newEntryAction->setShortcut(QKeySequence::New);
    connect(newEntryAction, &QAction::triggered, this, &MainWindow::addNewEntry);

    removeEntryAction = new QAction(tr("Remove entry"), this);
    removeEntryAction->setShortcut(QKeySequence::Delete);
    connect(removeEntryAction, &QAction::triggered, this, &MainWindow::removeEntry);

    demoAction = new QAction(tr("Demo entry"), this);
    connect(demoAction, &QAction::triggered, this, &MainWindow::addDemoEntry);

    printAction = new QAction(tr("Print Preview..."), this);
    connect(printAction, &QAction::triggered, this, &MainWindow::printPreview);

    zoomInAction = new QAction(tr("Zoom In"), this);
    zoomInAction->setShortcut(QKeySequence::ZoomIn);
    connect(zoomInAction, &QAction::triggered, this, &MainWindow::zoomIn);

    zoomOutAction = new QAction(tr("Zoom Out"), this);
    zoomOutAction->setShortcut(QKeySequence::ZoomOut);
    connect(zoomOutAction, &QAction::triggered, this, &MainWindow::zoomOut);

    ui->ganttView->leftView()->setContextMenuPolicy(Qt::CustomContextMenu);
    ui->ganttView->leftView()->addAction(newEntryAction);
    ui->ganttView->leftView()->addAction(removeEntryAction);

    QMenu *entryMenu = menuBar()->addMenu(tr("Entry"));
    entryMenu->addAction(newEntryAction);
    entryMenu->addAction(removeEntryAction);
    entryMenu->addSeparator();
    entryMenu->addAction(demoAction);
    entryMenu->addSeparator();
    entryMenu->addAction(printAction);

    QMenu *zoomMenu = menuBar()->addMenu(tr("Zoom"));
    zoomMenu->addAction(zoomInAction);
    zoomMenu->addAction(zoomOutAction);

    enableActions(QItemSelection());
}

void MainWindow::initItemDelegate()
{
    auto *delegate = new EntryDelegate(constraintModel, this);
    ui->ganttView->leftView()->setItemDelegate(delegate);
}

void MainWindow::initGrid()
{
    grid = new KDGantt::DateTimeGrid();
    grid->setDayWidth(dayWidth);
    ui->ganttView->setGrid(grid);
}

void MainWindow::showContextMenu(const QPoint &pos)
{
    if (!ui->ganttView->leftView()->indexAt(pos).isValid())
        ui->ganttView->selectionModel()->clearSelection();

    QMenu menu(ui->ganttView->leftView());
    menu.addAction(newEntryAction);
    menu.addAction(removeEntryAction);
    menu.exec(ui->ganttView->leftView()->viewport()->mapToGlobal(pos));
}

void MainWindow::enableActions(const QItemSelection &selected)
{
    if (selected.indexes().isEmpty()) {
        newEntryAction->setEnabled(true);
        removeEntryAction->setEnabled(false);
        return;
    }

    QModelIndex selectedIndex = selected.indexes()[0];

    if (model->data(model->index(selectedIndex.row(), 1)) == KDGantt::TypeEvent || model->data(model->index(selectedIndex.row(), 1)) == KDGantt::TypeTask) {
        newEntryAction->setEnabled(false);
        removeEntryAction->setEnabled(true);
        return;
    }

    newEntryAction->setEnabled(true);
    removeEntryAction->setEnabled(true);
}

void MainWindow::addNewEntry()
{
    QPointer<EntryDialog> dialog = new EntryDialog(model);
    dialog->setWindowTitle(tr("New Entry"));
    if (dialog->exec() == QDialog::Rejected || !dialog) {
        delete dialog;
        return;
    }

    QModelIndexList selectedIndexes = ui->ganttView->selectionModel()->selectedIndexes();
    const QModelIndex parent = selectedIndexes.value(0);

    if (!model->insertRow(model->rowCount(parent), parent))
        return;

    int row = model->rowCount(parent) - 1;
    if (row == 0 && parent.isValid())
        model->insertColumns(model->columnCount(parent), 5, parent);

    model->setData(model->index(row, 0, parent), dialog->name());
    model->setData(model->index(row, 1, parent), dialog->type());
    if (dialog->type() != KDGantt::TypeSummary) {
        model->setData(model->index(row, 2, parent), dialog->startDate(), KDGantt::StartTimeRole);
        model->setData(model->index(row, 3, parent), dialog->endDate(), KDGantt::EndTimeRole);
    }
    model->setData(model->index(row, 4, parent), dialog->completion());
    model->setData(model->index(row, 5, parent), dialog->legend());

    addConstraint(dialog->depends(), model->index(row, 0, parent));
    setReadOnly(model->index(row, 0, parent), dialog->readOnly());

    delete dialog;
}

void MainWindow::setReadOnly(const QModelIndex &index, bool readOnly)
{
    int row = index.row();
    const QModelIndex parent = index.parent();
    QStandardItem *item;

    item = model->itemFromIndex(model->index(row, 0, parent));
    item->setFlags(readOnly ? item->flags() & ~Qt::ItemIsEditable : item->flags() | Qt::ItemIsEditable);

    item = model->itemFromIndex(model->index(row, 1, parent));
    item->setFlags(readOnly ? item->flags() & ~Qt::ItemIsEditable : item->flags() | Qt::ItemIsEditable);

    item = model->itemFromIndex(model->index(row, 2, parent));
    item->setFlags(readOnly ? item->flags() & ~Qt::ItemIsEditable : item->flags() | Qt::ItemIsEditable);

    item = model->itemFromIndex(model->index(row, 3, parent));
    item->setFlags(readOnly ? item->flags() & ~Qt::ItemIsEditable : item->flags() | Qt::ItemIsEditable);

    item = model->itemFromIndex(model->index(row, 4, parent));
    item->setFlags(readOnly ? item->flags() & ~Qt::ItemIsEditable : item->flags() | Qt::ItemIsEditable);
}

void MainWindow::addConstraint(const QModelIndex &index1, const QModelIndex &index2)
{
    if (!index1.isValid() || !index2.isValid())
        return;

    KDGantt::Constraint c(index1, index2);
    ui->ganttView->constraintModel()->addConstraint(c);
}

void MainWindow::addConstraint(const QStandardItem *item1, const QStandardItem *item2)
{
    addConstraint(model->indexFromItem(item1), model->indexFromItem(item2));
}

void MainWindow::removeEntry()
{
    QModelIndexList selectedIndexes = ui->ganttView->selectionModel()->selectedIndexes();
    QModelIndex index = selectedIndexes.value(0);

    if (!index.isValid())
        return;

    model->removeRow(index.row(), index.parent());
}

class MyStandardItem : public QStandardItem
{
public:
    MyStandardItem(const QVariant &v)
        : QStandardItem()
    {
        setData(v, Qt::DisplayRole);
    }
    MyStandardItem(const QString &v)
        : QStandardItem()
    {
        setData(v, Qt::DisplayRole);
    }
    MyStandardItem(const QDateTime &dt, int role)
        : QStandardItem()
    {
        setData(QVariant::fromValue(dt), role);
    }
};

void MainWindow::addDemoEntry()
{
    QStandardItem *softwareRelease = new MyStandardItem(tr("Software Release"));
    QStandardItem *codeFreeze = new MyStandardItem(tr("Code Freeze"));
    codeFreeze->setData(KDGantt::TextPositionRole, KDGantt::StyleOptionGanttItem::Right);
    QStandardItem *packaging = new MyStandardItem(tr("Packaging"));
    QStandardItem *upload = new MyStandardItem(tr("Upload"));
    QStandardItem *testing = new MyStandardItem(tr("Testing"));
    QStandardItem *updateDocumentation = new MyStandardItem(tr("Update Documentation"));

    QDateTime now = QDateTime::currentDateTime();

    softwareRelease->appendRow(QList<QStandardItem *>()
                               << codeFreeze << new MyStandardItem(KDGantt::TypeEvent)
                               << new MyStandardItem(now, KDGantt::StartTimeRole));
    softwareRelease->appendRow(QList<QStandardItem *>()
                               << packaging << new MyStandardItem(KDGantt::TypeTask)
                               << new MyStandardItem(now.addDays(5), KDGantt::StartTimeRole)
                               << new MyStandardItem(now.addDays(10), KDGantt::EndTimeRole));
    softwareRelease->appendRow(QList<QStandardItem *>()
                               << upload << new MyStandardItem(KDGantt::TypeTask)
                               << new MyStandardItem(now.addDays(10).addSecs(2 * 60 * 60), KDGantt::StartTimeRole)
                               << new MyStandardItem(now.addDays(11), KDGantt::EndTimeRole));
    softwareRelease->appendRow(QList<QStandardItem *>()
                               << testing << new MyStandardItem(KDGantt::TypeTask)
                               << new MyStandardItem(now.addSecs(3 * 60 * 60), KDGantt::StartTimeRole)
                               << new MyStandardItem(now.addDays(5), KDGantt::EndTimeRole));
    softwareRelease->appendRow(QList<QStandardItem *>()
                               << updateDocumentation << new MyStandardItem(KDGantt::TypeTask)
                               << new MyStandardItem(now.addSecs(3 * 60 * 60), KDGantt::StartTimeRole)
                               << new MyStandardItem(now.addDays(3), KDGantt::EndTimeRole));

    model->appendRow(QList<QStandardItem *>()
                     << softwareRelease << new MyStandardItem(KDGantt::TypeSummary));

    addConstraint(codeFreeze, packaging);
    addConstraint(codeFreeze, testing);
    addConstraint(codeFreeze, updateDocumentation);
    addConstraint(packaging, upload);
    addConstraint(testing, packaging);
    addConstraint(updateDocumentation, packaging);
}

void MainWindow::zoomIn()
{
    dayWidth += 10;
    if (dayWidth > 400)
        grid->setScale(KDGantt::DateTimeGrid::ScaleHour);

    grid->setDayWidth(dayWidth);
}

void MainWindow::zoomOut()
{
    dayWidth -= 10;
    if (dayWidth < 10)
        dayWidth = 10;

    if (dayWidth <= 400)
        grid->setScale(KDGantt::DateTimeGrid::ScaleDay);

    grid->setDayWidth(dayWidth);
}

void MainWindow::printPreview()
{
    auto *preview = new QLabel(this, Qt::Window);
    preview->setAttribute(Qt::WA_DeleteOnClose);
    preview->setScaledContents(true);
    preview->setWindowTitle(tr("Print Preview"));
    QPixmap pix(1000, 300);
    pix.fill(Qt::white);
    {
        QPainter p(&pix);
        p.setRenderHints(QPainter::Antialiasing);
        ui->ganttView->print(&p, pix.rect());
    }
    preview->setPixmap(pix);
    preview->show();
}

void MainWindow::slotClicked(const QModelIndex &index)
{
    statusBar()->showMessage(tr("(%1,%2,_,%4) clicked")
                                 .arg(index.row())
                                 .arg(index.column())
                                 .arg(( quint64 )index.model()));
}
void MainWindow::slotDoubleClicked(const QModelIndex &index)
{
    statusBar()->showMessage(tr("(%1,%2,_,%4) qreal clicked")
                                 .arg(index.row())
                                 .arg(index.column())
                                 .arg(( quint64 )index.model()));
}

File: python/examples/apireview/mainwindow.py
#!/usr/bin/env python

#
# This file is part of the KD Chart library.
#
# SPDX-FileCopyrightText: 2019-2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
#
# SPDX-License-Identifier: MIT
#

''' MainWindow for the API Review Example '''

# pylint: disable=missing-function-docstring,missing-class-docstring

from PySide2.QtCore import Qt, QDateTime, QModelIndex
from PySide2.QtGui import QStandardItem, QStandardItemModel, QPixmap, QKeySequence, QPainter
from PySide2.QtWidgets import QMainWindow, QAction, QMenu, QLabel, QDialog, QWidget, QMenuBar
from PyKDChart.KDGantt import ConstraintModel, DateTimeGrid, StyleOptionGanttItem, Constraint
from PyKDChart import KDGantt

from ui_mainwindow import Ui_MainWindow
from entrydelegate import EntryDelegate
from entrydialog import EntryDialog

# pylint: disable=too-few-public-methods


class MyStandardItem(QStandardItem):
    def __init__(self, v, role=Qt.DisplayRole):
        super().__init__()
        self.setData(v, role)


class MainWindow(QMainWindow):
    # pylint: disable=too-many-instance-attributes
    def __init__(self, parent=None, flags=Qt.WindowFlags()):
        super().__init__(parent, flags)

        self.dayWidth = 70
        self.ui = Ui_MainWindow()  # pylint: disable=invalid-name
        self.ui.setupUi(self)

        self.initModel()
        self.initActions()
        self.initItemDelegate()
        self.initGrid()

        leftView = self.ui.ganttView.leftView()
        leftView.setColumnHidden(1, True)
        leftView.setColumnHidden(2, True)
        leftView.setColumnHidden(3, True)
        leftView.setColumnHidden(4, True)
        leftView.setColumnHidden(5, True)
        leftView.header().setStretchLastSection(True)

        self.ui.ganttView.leftView().customContextMenuRequested.connect(self.showContextMenu)
        self.ui.ganttView.selectionModel().selectionChanged.connect(self.enableActions)
        self.ui.ganttView.graphicsView().clicked.connect(self.slotClicked)
        self.ui.ganttView.graphicsView().qrealClicked.connect(self.slotDoubleClicked)

    def initModel(self):
        self.model = QStandardItemModel(0, 6, self)
        self.model.setHeaderData(0, Qt.Horizontal, "Task")
        self.ui.ganttView.setModel(self.model)

        self.l = QWidget(self)
        self.l.setWindowTitle("Legend")
        self.l.show()
        # self.l.setModel(self.model)

        self.constraintModel = ConstraintModel(self)
        self.ui.ganttView.setConstraintModel(self.constraintModel)

    def initActions(self):
        self.newEntryAction = QAction("New entry", self)
        self.newEntryAction.setShortcut(QKeySequence.New)
        self.newEntryAction.triggered.connect(self.addNewEntry)

        self.removeEntryAction = QAction("Remove entry", self)
        self.removeEntryAction.setShortcut(QKeySequence.Delete)
        self.removeEntryAction.triggered.connect(self.removeEntry)

        self.demoAction = QAction("Demo entry", self)
        self.demoAction.triggered.connect(self.addDemoEntry)

        self.printAction = QAction("Print Preview...", self)
        self.printAction.triggered.connect(self.printPreview)

        self.zoomInAction = QAction("Zoom In", self)
        self.zoomInAction.setShortcut(QKeySequence.ZoomIn)
        self.zoomInAction.triggered.connect(self.zoomIn)

        self.zoomOutAction = QAction("Zoom Out", self)
        self.zoomOutAction.setShortcut(QKeySequence.ZoomOut)
        self.zoomOutAction.triggered.connect(self.zoomOut)

        self.ui.ganttView.leftView().setContextMenuPolicy(Qt.CustomContextMenu)
        self.ui.ganttView.leftView().addAction(self.newEntryAction)
        self.ui.ganttView.leftView().addAction(self.removeEntryAction)

        menuBar = QMenuBar(self)
        self.setMenuBar(menuBar)
        entryMenu = menuBar.addMenu("Entry")
        entryMenu.addAction(self.newEntryAction)
        entryMenu.addAction(self.removeEntryAction)
        entryMenu.addSeparator()
        entryMenu.addAction(self.demoAction)
        entryMenu.addSeparator()
        entryMenu.addAction(self.printAction)

        zoomMenu = menuBar.addMenu("Zoom")
        zoomMenu.addAction(self.zoomInAction)
        zoomMenu.addAction(self.zoomOutAction)

        # self.enableActions(QItemSelection())

    def initItemDelegate(self):
        delegate = EntryDelegate(self.constraintModel, self)
        self.ui.ganttView.leftView().setItemDelegate(delegate)

    def initGrid(self):
        self.grid = DateTimeGrid()
        self.grid.setDayWidth(self.dayWidth)
        self.ui.ganttView.setGrid(self.grid)

    def showContextMenu(self, pos):
        if not self.ui.ganttView.leftView().indexAt(pos).isValid():
            self.ui.ganttView.selectionModel().clearSelection()

        menu = QMenu(self.ui.ganttView.leftView())
        menu.addAction(self.newEntryAction)
        menu.addAction(self.removeEntryAction)
        menu.exec_(self.ui.ganttView.leftView().viewport().mapToGlobal(pos))

    def enableActions(self, selected):
        if len(selected.indexes()) == 0:
            self.newEntryAction.setEnabled(True)
            self.removeEntryAction.setEnabled(False)
            return

        selectedIndex = selected.indexes()[0]
        dataType = self.model.data(self.model.index(selectedIndex.row(), 1))
        if dataType in [KDGantt.TypeEvent, KDGantt.TypeTask]:
            self.newEntryAction.setEnabled(False)
            self.removeEntryAction.setEnabled(True)
            return

        self.newEntryAction.setEnabled(True)
        self.removeEntryAction.setEnabled(True)

    def addNewEntry(self):
        dialog = EntryDialog(self.model)
        dialog.setWindowTitle("New Entry")
        if dialog.exec_() == QDialog.Rejected:
            dialog = None
            return

        selectedIndexes = self.ui.ganttView.selectionModel().selectedIndexes()
        if len(selectedIndexes) > 0:
            parent = selectedIndexes[0]
        else:
            parent = QModelIndex()

        if not self.model.insertRow(self.model.rowCount(parent), parent):
            return

        row = self.model.rowCount(parent) - 1
        if row == 0 and parent.isValid():
            self.model.insertColumns(self.model.columnCount(parent), 5, parent)

        self.model.setData(self.model.index(row, 0, parent), dialog.name())
        self.model.setData(self.model.index(row, 1, parent), dialog.type())
        if dialog.type() != KDGantt.TypeSummary:
            self.model.setData(self.model.index(row, 2, parent),
                               dialog.startDate(), KDGantt.StartTimeRole)
            self.model.setData(self.model.index(row, 3, parent),
                               dialog.endDate(), KDGantt.EndTimeRole)

        self.model.setData(self.model.index(
            row, 4, parent), dialog.completion())
        self.model.setData(self.model.index(row, 5, parent), dialog.legend())

        self.addConstraint(dialog.depends(), self.model.index(row, 0, parent))
        self.setReadOnly(self.model.index(row, 0, parent), dialog.readOnly())

        dialog = None

    def setReadOnly(self, index, readOnly):
        row = index.row()
        parent = index.parent()

        for column in range(0, 5):
            item = self.model.itemFromIndex(
                self.model.index(row, column, parent))
            flags = None
            if readOnly:
                flags = item.flags() & ~Qt.ItemIsEditable
            else:
                flags = item.flags() | Qt.ItemIsEditable
            item.setFlags(flags)

    def addConstraint(self, index1, index2):
        if not index1.isValid() or not index2.isValid():
            return

        c = Constraint(index1, index2)
        self.ui.ganttView.constraintModel().addConstraint(c)

    def addConstraintFromItem(self, item1, item2):
        self.addConstraint(self.model.indexFromItem(
            item1), self.model.indexFromItem(item2))

    def removeEntry(self):
        selectedIndexes = self.ui.ganttView.selectionModel().selectedIndexes()
        if len(selectedIndexes) > 0:
            index = selectedIndexes[0]
        else:
            index = QModelIndex()

        if not index.isValid():
            return

        self.model.removeRow(index.row(), index.parent())

    def addDemoEntry(self):
        softwareRelease = MyStandardItem("Software Release")
        codeFreeze = MyStandardItem("Code Freeze")
        codeFreeze.setData(KDGantt.TextPositionRole,
                           StyleOptionGanttItem.Right)
        packaging = MyStandardItem("Packaging")
        upload = MyStandardItem("Upload")
        testing = MyStandardItem("Testing")
        updateDocumentation = MyStandardItem("Update Documentation")

        now = QDateTime.currentDateTime()
        softwareRelease.appendRow([codeFreeze,
                                   MyStandardItem(KDGantt.TypeEvent),
                                   MyStandardItem(now, KDGantt.StartTimeRole)])
        softwareRelease.appendRow([packaging,
                                   MyStandardItem(KDGantt.TypeTask),
                                   MyStandardItem(now.addDays(
                                       5), KDGantt.StartTimeRole),
                                   MyStandardItem(now.addDays(10), KDGantt.EndTimeRole)])
        softwareRelease.appendRow([upload,
                                   MyStandardItem(KDGantt.TypeTask),
                                   MyStandardItem(now.addDays(10).addSecs(
                                       2 * 60 * 60), KDGantt.StartTimeRole),
                                   MyStandardItem(now.addDays(11), KDGantt.EndTimeRole)])
        softwareRelease.appendRow([testing,
                                   MyStandardItem(KDGantt.TypeTask),
                                   MyStandardItem(now.addSecs(
                                       3 * 60 * 60), KDGantt.StartTimeRole),
                                   MyStandardItem(now.addDays(5), KDGantt.EndTimeRole)])
        softwareRelease.appendRow([updateDocumentation,
                                   MyStandardItem(KDGantt.TypeTask),
                                   MyStandardItem(now.addSecs(
                                       3 * 60 * 60), KDGantt.StartTimeRole),
                                   MyStandardItem(now.addDays(3), KDGantt.EndTimeRole)])
        self.model.appendRow(
            [softwareRelease, MyStandardItem(KDGantt.TypeSummary)])
        self.addConstraintFromItem(codeFreeze, packaging)
        self.addConstraintFromItem(codeFreeze, testing)
        self.addConstraintFromItem(codeFreeze, updateDocumentation)
        self.addConstraintFromItem(packaging, upload)
        self.addConstraintFromItem(testing, packaging)
        self.addConstraintFromItem(updateDocumentation, packaging)

    def zoomIn(self):
        self.dayWidth += 10
        if self.dayWidth > 400:
            self.grid.setScale(DateTimeGrid.ScaleHour)
        self.grid.setDayWidth(self.dayWidth)

    def zoomOut(self):
        self.dayWidth -= 10

        self.dayWidth = max(self.dayWidth, 10)

        if self.dayWidth <= 400:
            self.grid.setScale(DateTimeGrid.ScaleDay)

        self.grid.setDayWidth(self.dayWidth)

    def printPreview(self):
        preview = QLabel(self, Qt.Window)
        preview.setAttribute(Qt.WA_DeleteOnClose)
        preview.setScaledContents(True)
        preview.setWindowTitle("Print Preview")
        pix = QPixmap(1000, 300)
        pix.fill(Qt.white)

        p = QPainter(pix)
        p.setRenderHints(QPainter.Antialiasing)
        self.ui.ganttView.print_(p, pix.rect())

        preview.setPixmap(pix)
        preview.show()

    def slotClicked(self, index):
        self.statusBar().showMessage("(%d,%d,_,%s) clicked" %
                                     (index.row(), index.column(), index.model()))

    def slotDoubleClicked(self, index):
        self.statusBar().showMessage("(%d,%d,_,%s) qreal clicked" %
                                     (index.row(), index.column(), index.model()))
